hydra-node/src/Hydra/Chain/Direct.hs

```diff
@@ -285,15 +294,100 @@ client tracer point queue headState callback nodeToClientV =
 
+-- | The on-chain head state is maintained in a mutually-recursive data-structure,
+-- pointing to the previous chain state; This allows to easily rewind the state
+-- to a point in the past when rolling backward due to a change of chain fork by
+-- our local cardano-node.
+data SomeOnChainHeadStateAt = SomeOnChainHeadStateAt
+  { currentOnChainHeadState :: SomeOnChainHeadState
+  , recordedAt :: RecordedAt
+  }
+  deriving (Eq, Show)
+
+-- | Records when a state was seen on-chain. 'AtStart' is used for states that
+-- simply exist out of any chain events (e.g. the 'Idle' state).
+data RecordedAt
+  = AtStart
+  | AtPoint ChainPoint SomeOnChainHeadStateAt
+  deriving (Eq, Show)
+
+data ChainSyncHandler m = ChainSyncHandler
+  { onRollForward :: Block -> m ()
+  , onRollBackward :: Point Block -> m ()
+  }
+
+newChainSyncHandler ::
+  forall m.
+  (MonadSTM m) =>
+  -- | Tracer for logging
+  Tracer m DirectChainLog ->
+  -- | Chain callback
+  (ChainEvent Tx -> m ()) ->
+  -- | On-chain head-state.
+  TVar m SomeOnChainHeadStateAt ->
+  -- | A chain-sync handler to use in a local-chain-sync client.
+  m (ChainSyncHandler m)
+newChainSyncHandler tracer callback headState = do
+  pure $
+    ChainSyncHandler
+      { onRollBackward
+      , onRollForward
+      }
+ where
+  onRollBackward :: Point Block -> m ()
+  onRollBackward point = do
+    traceWith tracer $ RolledBackward $ SomePoint point
+    (st, depth) <- rollback (fromConsensusPointHF point) <$> readTVarIO headState
+    atomically $ writeTVar headState st
+    callback (Rollback depth)
+
+  onRollForward :: Block -> m ()
+  onRollForward blk = do
+    let receivedTxs = toList $ getAlonzoTxs blk
+    onChainTxs <- reverse <$> atomically (foldM (withNextTx (blockPoint blk)) [] receivedTxs)
+    unless (null receivedTxs) $
+      traceWith tracer $
+        ReceivedTxs
+          { onChainTxs
+          , receivedTxs = map (\tx -> (getTxId tx, tx)) receivedTxs
+          }
+    mapM_ (callback . Observation) onChainTxs
+
+  withNextTx :: Point Block -> [OnChainTx Tx] -> ValidatedTx Era -> STM m [OnChainTx Tx]
+  withNextTx point observed (fromLedgerTx -> tx) = do
+    st <- readTVar headState
+    case observeSomeTx tx (currentOnChainHeadState st) of
+      Just (onChainTx, st') -> do
+        writeTVar headState $
+          SomeOnChainHeadStateAt
+            { currentOnChainHeadState = st'
+            , recordedAt = AtPoint (fromConsensusPointHF point) st
+            }
+        pure $ onChainTx : observed
+      Nothing ->
+        pure observed
+
+-- | Rewind some head state back to the first known state that is strictly
+-- before the provided 'ChainPoint'.
+rollback :: ChainPoint -> SomeOnChainHeadStateAt -> (SomeOnChainHeadStateAt, Word)
+rollback pt = backward 0
+ where
+  backward n st =
+    case recordedAt st of
+      AtStart ->
+        (st, n)
+      AtPoint at previous ->
+        if at <= pt
+          then (st, n)
+          else backward (succ n) previous
```

Comment:
My brain is fried, but we should be careful about this function and likely revisit it or carefully evaluate the consequences of it. While simple in appearance, this might subtly create a data-inconsistency between the chain-sync server and our client. 

Indeed, the server asks us to roll back to point X, and we actually rollback to a point **prior to** X. Thus, when the server start again streaming block, it'll do so from X+1, and so on. Thus, we end up with a (at least mental) gap between where we're at and where the server is at. 

I don't think this matter in practice, because we know that by construction, there is _no observable transaction_ between where we're at (i.e. the last known state) and where the server will continue streaming block from (i.e. the point of rollback). This would however definitely look weird if we started logging all blocks that we see; the gap would be visible. 

So, probably nothing, but maybe worth a comment and definitely worth revisiting with less caffeine and more sleep in the body.

Reply:
The whole concept also seems to rely on the fact take both layers are consistent in how many chain states they have in their respective lists. While this (seemingly) does work, it is quite fragile and I would propose to have instead the more complex callback, but single state ADR18 (in a follow-up iteration of course)

Reply:
On the one hand, there's really nothing forcing us to represent states on both side in the same manner, we just happen to use the same approach because it's the most obvious one and simple one.

On the other hand, I do agree with you and it makes me uneasy to have pretty much the same logic on both ends.